File: LK.ml
open FOL

type judgement = Judge of formula list * formula list

let show_judgement (Judge(assms, prop)) =
  "[" ^ show_formulas (List.rev assms) ^ "] |- [" ^ show_formulas prop ^ "]"
let show_judgements js = String.concat "\n" (List.map show_judgement js)

let formulize : judgement -> formula =
  fun (Judge(assms, props)) ->
  Then (List.fold_left (fun x y -> And(x,y)) Top assms,
        List.fold_left (fun x y -> Or(x,y)) Bottom props)

type assmIndex = string

type rule =
  | I | Cut of formula
  | AndL1 | AndL2 | AndR
  | OrL | OrR1 | OrR2
  | ImpL | ImpR
  | BottomL | TopR
  | ForallL of term | ForallR of ident
  | ExistL of ident | ExistR of term
  | WL | WR
  | CL | CR
  | PL of int | PR of int

let show_rule = function
  | I -> Printf.sprintf "I"
  | Cut(f) -> Printf.sprintf "Cut(%s)" (show_formula f)
  | AndL1 -> Printf.sprintf "AndL1"
  | AndL2 -> Printf.sprintf "AndL2"
  | AndR -> Printf.sprintf "AndR"
  | OrL -> Printf.sprintf "OrL"
  | OrR1 -> Printf.sprintf "OrR1"
  | OrR2 -> Printf.sprintf "OrR2"
  | ImpL -> Printf.sprintf "ImpL"
  | ImpR -> Printf.sprintf "ImpR"
  | BottomL -> Printf.sprintf "BottomL"
  | TopR -> Printf.sprintf "TopR"
  | ForallL(t) -> Printf.sprintf "ForallL(%s)" (show_term t)
  | ForallR(x) -> Printf.sprintf "ForallR(%S)" x
  | ExistL(x) -> Printf.sprintf "ExistL(%S)" x
  | ExistR(t) -> Printf.sprintf "ExistR(%s)" (show_term t)
  | WL -> Printf.sprintf "WL"
  | WR -> Printf.sprintf "WR"
  | CL -> Printf.sprintf "CL"
  | CR -> Printf.sprintf "CR"
  | PL(i) -> Printf.sprintf "PL(%d)" i
  | PR(i) -> Printf.sprintf "PR(%d)" i
let show_rules rules = String.concat ", " (List.map show_rule rules)

File: lk.pl
:- module(lk,[judgement/1,assmIndex/1,rule/1,judge/3]).
:- use_module(fol).
judgement(Fs1 ⊦ Fs2) :- maplist(formula,Fs1),maplist(formula,Fs2).
assmIndex(I) :- string(I).
rule(i). rule(cut(F)) :- formula(F).
rule(andL1). rule(andL2). rule(andR).
rule(orL). rule(orR1). rule(orR2).
rule(impL). rule(impR). rule(bottomL). rule(topR).
rule(forallL(T)) :- term(T). rule(forallR(I)) :- ident(I).
rule(existL(I)) :- ident(I). rule(existR(T)) :- term(T).
rule(wL). rule(wR). rule(cL). rule(cR).
rule(pL(I)) :- integer(I). rule(pR(I)) :- integer(I).

%judge(Rs,Js,_) :- writeln(judge(Rs;Js)),fail.
judge([i|R],[A⊦A|J],J_) :- judge(R,J,J_).
judge([cut(F)|R],[A⊦P|J],J_) :- judge(R,[A⊦[F|P],[F|A]⊦P|J],J_).
judge([andL1|R],[[and(F,_)|A]⊦P|J],J_) :- judge(R,[[F|A]⊦P|J],J_).
judge([andL2|R],[[and(_,F)|A]⊦P|J],J_) :- judge(R,[[F|A]⊦P|J],J_).
judge([andR|R],[A⊦[and(F1,F2)|P]|J],J_) :- judge(R,[A⊦[F1|P],A⊦[F2|P]|J],J_).
judge([orL|R],[[or(F1,F2)|A]⊦P|J],J_) :- judge(R,[[F1|A]⊦P,[F2|A]⊦P|J],J_).
judge([orR1|R],[A⊦[or(F,_)|P]|J],J_) :- judge(R,[A⊦[F|P]|J],J_).
judge([orR2|R],[A⊦[or(_,F)|P]|J],J_) :- judge(R,[A⊦[F|P]|J],J_).
judge([impL|R],[[(F1==>F2)|A]⊦P|J],J_) :- judge(R,[A⊦[F1|P],[F2|A]⊦P|J],J_).
judge([impR|R],[A⊦[(F1==>F2)|P]|J],J_) :- judge(R,[[F1|A]⊦[F2|P]|J],J_).
judge([bottomL|R],[[bottom|_]⊦_|J],J_) :- judge(R,J,J_).
judge([topR|R],[_⊦[top|_]|J],J_) :- judge(R,J,J_).
judge([forallL(T)|R],[[forall(X,F)|A]⊦P|J],J_) :- substFormula(X,T,F,F_),judge(R,[[F_|A]⊦P|J],J_).
judge([forallR(Y)|R],[A⊦[forall(X,F)|P]|J],J_) :- substFormula(X,!Y,F,F_),judge(R,[A⊦[F_|P]|J],J_).
judge([existL(Y)|R],[[exist(X,F)|A]⊦P|J],J_) :- substFormula(X,!Y,F,F_),judge(R,[[F_|A]⊦P|J],J_).
judge([existR(T)|R],[A⊦[exist(X,F)|P]|J],J_) :- substFormula(X,T,F,F_),judge(R,[A⊦[F_|P]|J],J_).
judge([wL|R],[[_|A]⊦P|J],J_) :- judge(R,[A⊦P|J],J_).
judge([wR|R],[A⊦[_|P]|J],J_) :- judge(R,[A⊦P|J],J_).
judge([cL|R],[[F|A]⊦P|J],J_) :- judge(R,[[F,F|A]⊦P|J],J_).
judge([cR|R],[A⊦[F|P]|J],J_) :- judge(R,[A⊦[F,F|P]|J],J_).
judge([pL(K)|R],[A⊦P|J],J_) :- length(A,L),K<L,nth0(K,A,Ak,K_),judge(R,[[Ak|K_]⊦P|J],J_).
judge([pR(K)|R],[A⊦P|J],J_) :- length(P,L),K<L,nth0(K,P,Pk,P_),judge(R,[A⊦[Pk|P_]|J],J_).
judge([],J,J) :- !.
judge([R|_],J,_) :- throw(goalError(R,J)).
